File: app/src/main/java/ru/michanic/mymot/UI/Adapters/SearchMainAdapter.java
package ru.michanic.mymot.UI.Adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.gson.internal.LinkedTreeMap;
import com.squareup.picasso.Picasso;

import java.util.List;

import ru.michanic.mymot.Models.Advert;
import ru.michanic.mymot.Protocols.ClickListener;
import ru.michanic.mymot.Protocols.Const;
import ru.michanic.mymot.R;
import ru.michanic.mymot.UI.Cells.SearchMainCell;

public class SearchMainAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    Context context;
    List<Advert> adverts;
    ClickListener clickListener;

    public SearchMainAdapter(Context context, List<Advert> adverts, ClickListener clickListener) {
        this.context = context;
        this.adverts = adverts;
        this.clickListener = clickListener;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        View view = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.cell_advert_main, viewGroup, false);
        return new SearchMainCell(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, final int position) {
        if (holder instanceof SearchMainCell) {
            ((SearchMainCell) holder).fillWithAdvert(adverts.get(position));
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    clickListener.onClick(0, position);
                }
            });
        } else if (holder instanceof HeaderViewHolder) {
            ((HeaderViewHolder) holder).setTitle("Заголовок");
        }
    }

    @Override
    public int getItemCount() {
        return adverts.size();
    }


    class HeaderViewHolder extends RecyclerView.ViewHolder {
        public View View;
        private final TextView headerTitle;

        public HeaderViewHolder(View itemView) {
            super(itemView);
            View = itemView;
            // Добавьте свои компоненты ui здесь, как показано ниже
            headerTitle = (TextView) View.findViewById(R.id.headerTitle);

        }

        public void setTitle(String title) {
            headerTitle.setText(title);
        }
    }

}

File: app/src/main/java/ru/michanic/mymot/Kotlin/UI/Adapters/SearchMainAdapter.kt
package ru.michanic.mymot.Kotlin.UI.Adapters

import android.content.Context
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import ru.michanic.mymot.Kotlin.Models.Advert
import ru.michanic.mymot.Kotlin.Protocols.ClickListener
import ru.michanic.mymot.R
import ru.michanic.mymot.Kotlin.UI.Cells.SearchMainCell

class SearchMainAdapter(
    var context: Context,
    var adverts: List<Advert>,
    var clickListener: ClickListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val view = LayoutInflater.from(viewGroup.context)
            .inflate(R.layout.cell_advert_main, viewGroup, false)
        return SearchMainCell(view)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is SearchMainCell) {
            holder.fillWithAdvert(adverts[position])
            holder.itemView.setOnClickListener { clickListener.onClick(0, position) }
        } else if (holder is HeaderViewHolder) {
            holder.setTitle("Заголовок")
        }
    }

    override fun getItemCount(): Int {
        return adverts.size
    }

    internal inner class HeaderViewHolder(var View: View) : RecyclerView.ViewHolder(
        View
    ) {
        private val headerTitle: TextView = View.findViewById<View>(R.id.headerTitle) as TextView

        fun setTitle(title: String?) {
            headerTitle.text = title
        }
    }
}